File: strategy/generator/exit/ast.py
from dataclasses import dataclass

from core.models.parameter import (
    CategoricalParameter,
    Parameter,
    StaticParameter,
)
from core.models.smooth import SmoothATR
from core.models.source import SourceType

from .base import Exit, ExitType


@dataclass(frozen=True)
class AstExit(Exit):
    type: ExitType = ExitType.Ast
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    smooth_atr: Parameter = CategoricalParameter(SmoothATR)
    period_atr: Parameter = StaticParameter(12.0)
    factor: Parameter = StaticParameter(3.0)

File: ta_lib/strategies/exit/src/ast.rs
use base::prelude::*;
use core::prelude::*;
use trend::ast;

pub struct AstExit {
    source_type: SourceType,
    atr_period: usize,
    factor: f32,
}

impl AstExit {
    pub fn new(source_type: SourceType, atr_period: f32, factor: f32) -> Self {
        Self {
            source_type,
            atr_period: atr_period as usize,
            factor,
        }
    }
}

impl Exit for AstExit {
    fn lookback(&self) -> usize {
        self.atr_period
    }

    fn evaluate(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let (direction, _) = ast(
            &data.source(self.source_type),
            &data.atr(self.atr_period),
            self.factor,
        );

        (direction.cross_over(&ZERO), direction.cross_under(&ZERO))
    }
}
